src/OpenLoco/src/Windows/Construction/StationTab.cpp

```diff
@@ -211,20 +216,117 @@ namespace OpenLoco::Ui::Windows::Construction::Station
-        regs.ax = x;
-        regs.bx = y;
-        call(0x0049E421, regs);
+        size_t i = 0;
+        for (const auto& pos : range)
+        {
+            _mapSelectedTiles[i++] = pos;
+        }
+
+        _mapSelectedTiles[i].x = -1;
+        mapInvalidateMapSelectionTiles();
     }
 
+    static std::optional<GameCommands::AirportPlacementArgs> getAirportPlacementArgsFromCursor(const int16_t x, const int16_t y);
+
     static loco_global<World::Pos2, 0x001135F7C> _1135F7C;
     static loco_global<World::Pos2, 0x001135F80> _1135F90;
 
+    // 0x004A4F3B
+    static void onToolUpdateAirport(const Ui::Point& mousePos)
+    {
+        World::TileManager::mapInvalidateMapSelectionTiles();
+        Input::resetMapSelectionFlag(Input::MapSelectionFlags::enable | Input::MapSelectionFlags::enableConstruct | Input::MapSelectionFlags::enableConstructionArrow);
+        const auto args = getAirportPlacementArgsFromCursor(mousePos.x, mousePos.y);
+        if (!args.has_value())
+        {
+            // common for all tool updates move to parent func.
+            removeConstructionGhosts();
+            if (_stationCost != 0x80000000U)
+            {
+                _stationCost = 0x80000000U;
+                Ui::WindowManager::invalidate(Ui::WindowType::construction);
+            }
+            return;
+        }
+
+        Input::setMapSelectionFlags(Input::MapSelectionFlags::enableConstruct | Input::MapSelectionFlags::enableConstructionArrow);
+        Input::resetMapSelectionFlag(Input::MapSelectionFlags::unk_03);
+        _constructionArrowDirection = args->rotation;
+        _constructionArrowPos = args->pos;
+
+        setMapSelectedTilesFromRange(World::TilePosRangeView(World::TilePos2(*_1135F7C), World::TilePos2(*_1135F90)));
+
+        if (_byte_522096 & (1U << 3))
+        {
+            if (*_stationGhostPos == args->pos && *_stationGhostRotation == args->rotation && *_stationGhostTypeDockAirport == args->type)
+            {
+                return;
+            }
+            removeConstructionGhosts();
+        }
+
+        _stationGhostPos = args->pos;
+        _stationGhostRotation = args->rotation;
+        _stationGhostTypeDockAirport = args->type;
+        _stationGhostType = 0x8000; // Why??
+
+        const auto cost = GameCommands::doCommand(*args, GameCommands::Flags::apply | GameCommands::Flags::flag_3 | GameCommands::Flags::flag_5 | GameCommands::Flags::flag_6);
+
+        _stationCost = cost;
+
+        Ui::WindowManager::invalidate(Ui::WindowType::construction);
+
+        if (cost == GameCommands::FAILURE)
+        {
+            return;
+        }
+
+        _byte_522096 = _byte_522096 | (1U << 3);
+        Input::setMapSelectionFlags(Input::MapSelectionFlags::catchmentArea);
+        _constructingStationId = _lastConstructedGhostStationId;
+
+        auto* station = _lastConstructedGhostStationId != 0xFFFFFFFFU ? StationManager::get(static_cast<StationId>(*_lastConstructedGhostStationId)) : nullptr;
```

Comment:
_lastConstructedGhostStationId is named wrong its the station that the ghost will join to. 0xFFFFFFFFU if it will be a new station.